letsencrypt/client/ddns/ddns.py

```diff
@@ -0,0 +1,85 @@
+"""generic DDNS authenticator (RFC 2136)"""
+
+import logging
+import subprocess
+
+import zope.interface
+
+from letsencrypt.client import challenge_util
+from letsencrypt.client import interfaces
+
+
+class DDNS(object):
+    """DDNS authenticator.
+
+    """
+    zope.interface.implements(interfaces.IAuthenticator)
+
+    def get_chall_pref(self, unused_domain):  # pylint: disable=no-self-use
+        """Return list of challenge preferences."""
+        return ["dns"]
+
+    def perform(self, chall_list):  # pylint: disable=no-self-use
+        """Perform the configuration related challenge.
+
+        :param list chall_list: List of challenges to be
+            fulfilled by configurator.
+
+        :returns: list of responses. All responses are returned in the same
+            order as received by the perform function. A None response
+            indicates the challenge was not performed.
+        :rtype: list
+
+        """
+        responses = []
+
+        for chall in chall_list:
+            if isinstance(chall, challenge_util.DnsChall):
+                try:
+                    nsupdate("add", chall.domain, chall.token)
+                except subprocess.CalledProcessError:
+                    responses.append(None)
+                else:
+                    responses.append({"type": "dns"})
+            else:
+                responses.append(None)
+
+        return responses
+
+    def cleanup(self, chall_list):  # pylint: disable=no-self-use
+        """Revert all challenges."""
+        for chall in chall_list:
+            if isinstance(chall, challenge_util.DnsChall):
+                nsupdate("del", chall.domain, chall.token)
+
+
+# TODO: make this user-configurable.
+NSUPDATE_CMD = "nsupdate -k le-nsupdate.key"
+
+
+def nsupdate(action, domain, token):
+    """Invoke the nsupdate commandline tool to send a single DNS update
```

Comment:
How about http://www.dnspython.org/? Example `nsupdate`-like tool: https://gist.github.com/pklaus/4619865 that shows how to use `dns.update` module.